File: day9/day9.py
from typing import List, Tuple

with open("input.txt", "r") as file:
	input = file.read()

input = input.strip()

files: List[Tuple[int|None, int]] = []

id = 0
for index, digit in enumerate(input):
	digit = int(digit)

	if index % 2 == 0:
		files.append((id, digit))
		id += 1
	else:
		files.append((None, digit))

files_sorted = files

for x in range(len(files)-1, -1, -1):
	outer_file = files[x]
	outer_file_id, outer_file_count = outer_file

	if outer_file_id == None:
		continue

	for y, (inner_file_id, inner_file_count) in enumerate(files[:x]):
		if inner_file_count < outer_file_count or inner_file_id != None:
			continue

		files_sorted[x] = (None, outer_file_count)

		files_sorted[y] = outer_file

		if inner_file_count > outer_file_count:
			files_sorted.insert(y+1, (None, inner_file_count - outer_file_count))	

		break

flattened: List[int|None] = []
	
for x, (file_id, file_count) in enumerate(files_sorted):
	for y in range(file_count):
		flattened.append(file_id)

answer = 0

for x, y in enumerate(flattened):
	if y == None:
		continue

	answer += x * y

print(answer)

File: day9/day9.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func isSpaces(file []int) bool {
	for _, i := range file {
		if i != -1 {
			return false
		}
	}
	return true
}

func main() {
	inputFile, _ := os.ReadFile("input.txt")
	input := string(inputFile)

	files := [][]int{}

	id := 0

	for i, char := range strings.Trim(input, "\n") {
		if char != '0' {
			files = append(files, []int{})
		}

		if i%2 == 0 {
			for j := 0; j < int(char-'0'); j++ {
				files[len(files)-1] = append(files[len(files)-1], id)
			}
			id++
		} else {
			for j := 0; j < int(char-'0'); j++ {
				files[len(files)-1] = append(files[len(files)-1], -1)
			}
		}
	}

	fmt.Println(files)

	for filledFileIndex := len(files) - 1; filledFileIndex >= 0; filledFileIndex-- {
		filledFile := files[filledFileIndex]

		if isSpaces(filledFile) {
			continue
		}

		for emptyFileIndex, emptyFile := range files {
			if emptyFileIndex >= filledFileIndex {
				break
			}

			if !isSpaces(emptyFile) || len(filledFile) > len(emptyFile) {
				continue
			}

			files[emptyFileIndex] = filledFile

			justFilledExtraSpaces := []int{}
			for i := 0; i < len(emptyFile)-len(filledFile); i++ {
				justFilledExtraSpaces = append(justFilledExtraSpaces, -1)
			}

			justRemovedExtraSpaces := []int{}
			for i := 0; i < len(filledFile); i++ {
				justRemovedExtraSpaces = append(justRemovedExtraSpaces, -1)
			}

			files[filledFileIndex] = justRemovedExtraSpaces

			if len(justFilledExtraSpaces) > 0 {
				files = append(files[:emptyFileIndex+1], append([][]int{justFilledExtraSpaces}, files[emptyFileIndex+1:]...)...)
			}

			fmt.Println(files)

			break
		}
	}

	flattened := []int{}
	for _, file := range files {
		for _, block := range file {
			flattened = append(flattened, block)
		}
	}

	answer := 0

	for i, block := range flattened {
		if block == -1 {
			continue
		}
		answer += i * block
	}

	fmt.Println(answer)
}
